### PR TITLE
Bugfix/MTM-53267_extending_user_context_description

### PR DESCRIPTION
**Pull request by tsvet**
This PR is in the context of https://cumulocity.atlassian.net/browse/MTM-53267 . It adds more information to the Microservice SDK part regarding the user and tenant scope and their usage.

### CODE REVIEW

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 1]
```diff
@@ -30,40 +30,78 @@ Annotation | Description
 
 ### Context support {#context-support}
 
-It is described below the context support as utility tool for the user management described in [General aspects](/microservice-sdk/general-aspects) of microservices in {{< product-c8y-iot >}}.
+The context support is covered by the annotation `@EnableContextSupport`. It allows for choosing between `@TenantScope` and `@UserScope` which is related to the user management of microservices, as described in [General aspects](/microservice-sdk/general-aspects) in {{< product-c8y-iot >}}.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:33]:
```suggestion
The context support is covered by the annotation `@EnableContextSupport`. It allows to choose between `@TenantScope` and `@UserScope` which is related to the user management of microservices, as described in [General aspects](/microservice-sdk/general-aspects) in {{< product-c8y-iot >}}.
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 2]
```diff
@@ -30,40 +30,78 @@ Annotation | Description
...
 
-`@UserScope` and `@TenantScope` at type level annotation indicate that a bean created from class will be created in the scope defined. The user scope implies using tenant platform user credentials for platform calls. The tenant scope implies using service user credentials.
+Within the tenant scope, the service account credentials are used for the communication with the platform, while within the user scope the credentials of the authenticated user sending the request are used.
 
-An example of injecting a bean into the tenant scope is available in the platform API module as follows:
+#### Tenant scope  
+To create a bean in the tenant scope, you can use the annotation `@TenantScope`, as in the following code snippet.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:38]:
```suggestion
To create a bean in the tenant scope, use the annotation `@TenantScope`, as in the following example. 
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 3]
```diff
@@ -30,40 +30,78 @@ Annotation | Description
...
-    return platform.getEventApi();
-}  
+@Bean(name = "tenantMeasurementApi")
+  public MeasurementApi getMeasurementApi() throws SDKException {
+    return delegate.getMeasurementApi();
+  }
+```
+If you then use the bean `tenantMeasurementApi`, the context is automatically set to the `@TenantScope` and the service user credentials are utilized.
+
+By default, the beans are created in the tenant scope and use the service user to communicate with the platform. This behaviour is also enabled when using the respective annotation for the tenant context.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:49]:
```suggestion
By default, the beans are created in the tenant scope and use the service user to communicate with the platform. This behavior is also enabled when using the respective annotation for the tenant context. 
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 4]
```diff
@@ -30,40 +30,78 @@ Annotation | Description
...
+  public MeasurementApi getMeasurementApi() throws SDKException {
+    return delegate.getMeasurementApi();
+  }
+```
+If you then use the bean `tenantMeasurementApi`, the context is automatically set to the `@TenantScope` and the service user credentials are utilized.
+
+By default, the beans are created in the tenant scope and use the service user to communicate with the platform. This behaviour is also enabled when using the respective annotation for the tenant context. 
+
+There are predefined beans both in the `@TenantScope` and `@UserScope`. 
+The name of a bean in the tenant scope consists of the prefix `"tenant"` and the name of the respective API. Thus, to use MeasurementApi in the @TenantScope, you can specify @Qualifier("tenantMeasurementApi"), as shown in the example below. As the tenant scope is the default context for the created beans, the annotation can also be omitted. Therefore, the following two excerpts are equivalent and both suggest that the service user credentials will be used for the communication with the platform.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:52]:
```suggestion
The name of a bean in the tenant scope consists of the prefix "tenant" and the name of the respective API. Thus, to use the Measurement API in the tenant scope, you can specify `@Qualifier("tenantMeasurementApi")`, as shown in the example below. As the tenant scope is the default context for the created beans, the annotation can also be omitted. Therefore, the following two excerpts are equivalent and both suggest that the service user credentials will be used for the communication with the platform. 
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 5]
```diff
@@ -30,40 +30,78 @@ Annotation | Description
...
+@Autowired
+private MeasurementApi measurementApi;
 ```
+In both cases, beans within the tenant scope will be auto-wired. 
 
-A sample utilization of the bean can be as follows:
+#### User scope
+In certain situations the microservice should not use the service user credentials but the credentials of the user sending the request. 
+
+To create a bean from the same class in the user scope, you should specify `@UserScope`, as in the snippet below.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md]:
```suggestion
To create a bean from the same class in the user scope, specify `@UserScope`, as in the example below. 
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 6]
```diff
@@ -30,40 +30,78 @@ Annotation | Description
...
+
+```java
+@UserScope
+@Bean(name = "userMeasurementApi")
+public MeasurementApi getMeasurementApi() throws SDKException {
+  return delegate.getMeasurementApi();
+}
+```
+
+To use `@UserScope`, you have to use the respective annotations enabling it. Analogously to the tenant scope case, there are predefined beans in the user scope. The name of such beans consists of the prefix `"user"` and the name of the API. An example of auto-wiring a bean of the `@UserScope` is given below.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md]:
```suggestion
To use `@UserScope`, you must use the respective annotations enabling it. Analogously to the tenant scope case, there are predefined beans in the user scope. The name of such beans consists of the prefix "user" and the name of the API. An example of auto-wiring a bean of the `@UserScope` is given below.
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 7]
```diff
@@ -30,40 +30,78 @@ Annotation | Description
...
+
+```java
+@UserScope
+@Bean(name = "userMeasurementApi")
+public MeasurementApi getMeasurementApi() throws SDKException {
+  return delegate.getMeasurementApi();
+}
+```
+
+Analogously to the tenant scope case, there are predefined beans in the user scope. The name of such beans consists of the prefix `"user"` and the name of the API. An example of auto-wiring a bean of the `@UserScope` is given below.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:80]:
```suggestion
Analogously to the tenant scope case, there are predefined beans in the user scope. The name of such beans consists of the prefix "user" and the name of the API. An example of auto-wiring a bean of the `@UserScope` is given below.
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 8]
```diff
@@ -30,40 +30,78 @@ Annotation | Description
...
 
 ```java
 @Autowired
-private PlatformProperties platformProperties;
+@Qualifier("userMeasurementApi")
+private MeasurementApi measurementApi;
+```
+
+Within the user scope, the created beans use the credentials of the authenticated user sending the request instead of the default service user for the commication with the platform.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:88]:
```suggestion
Within the user scope, the created beans use the credentials of the authenticated user sending the request instead of the default service user for the communication with the platform. 
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 9]
```diff
@@ -30,40 +30,78 @@ Annotation | Description
...
 @Autowired
-private PlatformProperties platformProperties;
+@Qualifier("userMeasurementApi")
+private MeasurementApi measurementApi;
+```
+
+Within the user scope, the created beans use the credentials of the authenticated user sending the request instead of the default service user for the commication with the platform. 
+
+#### Setting the credentials explicitly
+We can explicitly set the credentials to be used through ContextService. To use the credentials of the user sending the request, we have to use `ContextService<UserCredentials>`. Analogously, `ContextService<MicroserviceCredentials>` can be used for service user credentials.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:91]:
```suggestion
You can explicitly set the credentials to be used through `ContextService`. To use the credentials of the user sending the request, use `ContextService<UserCredentials>`. Analogously, `ContextService<MicroserviceCredentials>` can be used for service user credentials. 
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 10]
```diff
@@ -30,40 +30,78 @@ Annotation | Description
...
-private PlatformProperties platformProperties;
+@Qualifier("userMeasurementApi")
+private MeasurementApi measurementApi;
+```
+
+Within the user scope, the created beans use the credentials of the authenticated user sending the request instead of the default service user for the commication with the platform. 
+
+#### Setting the credentials explicitly
+We can explicitly set the credentials to be used through ContextService. To use the credentials of the user sending the request, we have to use `ContextService<UserCredentials>`. Analogously, `ContextService<MicroserviceCredentials>` can be used for service user credentials. 
+An example on how to use ContextService is given below.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:92]:
```suggestion
An example on how to use `ContextService` is given below.
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 11]
```diff
@@ -30,40 +30,78 @@ Annotation | Description
...
+@Autowired
+private MeasurementApi measurementApi;
 ```
+In both cases, beans within the tenant scope will be auto-wired. 
 
-A sample utilization of the bean can be as follows:
+#### User scope
+In certain situations the microservice should not use the service user credentials but the credentials of the user sending the request. 
+
+To create a bean in the user scope, you should specify `@UserScope`, as in the snippet below.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:70]:
```suggestion
To create a bean in the user scope, specify `@UserScope`, as in the example below. 
```

**Comment by BeateRixen** [comment]
> @BeateRixen, thanks for the comments. I wanted to check the technical part first with the colleagues from the Platform management team, as the text might undergo some changes. Thats why I set the PR to draft, but I will try to keep track of the changes you requested already, as well.

Sorry @tsvet , I did not realize the draft status :-). Don´t worry, I will do another review when you are done with your changes.

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 12]
```diff
@@ -30,40 +30,78 @@ Annotation | Description
...
 @TenantScope
-public EventApi eventApi (Platform platform) throws SDKException {
-    return platform.getEventApi();
-}  
+@Bean(name = "tenantMeasurementApi")
+public MeasurementApi getMeasurementApi() throws SDKException {
+  return delegate.getMeasurementApi();
+}
+```
+If you then use the bean `tenantMeasurementApi`, the context is automatically set to the `@TenantScope` and the service user credentials are utilized.
```
Comment (jochenendres) [inline content/microservice-sdk/java-bundle/developing-microservice.md:47]: What you mean with "automatically set to the @TenantScope"?
This annotation seems to be set explicitly in the example!?

Comment (tsvet) [reply content/microservice-sdk/java-bundle/developing-microservice.md:47]: That the bean is in the tenant scope. I will try to rephrase it to make it clearer

**Comment by tsvet** [comment]
> > @BeateRixen, thanks for the comments. I wanted to check the technical part first with the colleagues from the Platform management team, as the text might undergo some changes. Thats why I set the PR to draft, but I will try to keep track of the changes you requested already, as well.
> 
> Sorry @tsvet , I did not realize the draft status :-). Don´t worry, I will do another review when you are done with your changes.

@BeateRixen , no problem at all. You were simply too fast :) I tried to add your suggestions where they still apply with my last commit. If you could check again at the end, it would be great.

File: content/microservice-sdk/java-bundle/developing-microservice.md (hunk 12 above)
Comment (tsvet) [reply content/microservice-sdk/java-bundle/developing-microservice.md]: I have removed "automatically".

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 13]
```diff
@@ -30,40 +30,79 @@ Annotation | Description
 
 ### Context support {#context-support}
 
-It is described below the context support as utility tool for the user management described in [General aspects](/microservice-sdk/general-aspects) of microservices in {{< product-c8y-iot >}}.
+The context support is covered by the annotation `@EnableContextSupport`. It allows to choose between `@TenantScope` and `@UserScope` which is related to the user management of microservices, as described in [General aspects](/microservice-sdk/general-aspects) in {{< product-c8y-iot >}}.
 
-`@UserScope` and `@TenantScope` at type level annotation indicate that a bean created from class will be created in the scope defined. The user scope implies using tenant platform user credentials for platform calls. The tenant scope implies using service user credentials.
+Each microservice has a service user which can be used for the interaction with the platform. The roles associated with this user are specified in the manifest. 
+Within the tenant scope, the credentials of this service user are used for the communication with the platform, while within the user scope the credentials of the authenticated user sending the request are used.
```
Comment (kaczynskid-sag) [inline content/microservice-sdk/java-bundle/developing-microservice.md:36]:
```suggestion
Within the tenant scope, the credentials of this service user are used for the communication with the platform, while within the user scope the credentials of the authenticated user sending the request to the microservice are used.
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 14]
```diff
@@ -30,40 +30,79 @@ Annotation | Description
...
 
 ```java
 @TenantScope
-public EventApi eventApi (Platform platform) throws SDKException {
-    return platform.getEventApi();
-}  
+@Bean(name = "tenantMeasurementApi")
+public MeasurementApi getMeasurementApi() throws SDKException {
+  return delegate.getMeasurementApi();
+}
```
Comment (kaczynskid-sag) [inline content/microservice-sdk/java-bundle/developing-microservice.md:46]: This is not a good example as it's not self contained due to unknown reference to `delegate` variable. Also the `name` attribute of the `@Bean` annotation is unnecessary in this context.

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 15]
```diff
@@ -30,40 +30,79 @@ Annotation | Description
...
+```
+If you then use the bean `tenantMeasurementApi`, the context is set to `@TenantScope` and the service user credentials are utilized.
```
Comment (kaczynskid-sag) [inline content/microservice-sdk/java-bundle/developing-microservice.md:48]: The context is set with `ContextService`. Using the scoped bean requires the context to be set upfront. It is not true, that using the bean will set the context.

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 16]
```diff
@@ -30,40 +30,79 @@ Annotation | Description
...
+
+By default, the beans are created in the tenant scope and use the service user to communicate with the platform. This behavior is also enabled when using the respective annotation for the tenant context.
```
Comment (kaczynskid-sag) [inline content/microservice-sdk/java-bundle/developing-microservice.md:50]: By default beans are created in `singleton` scope in Spring. The default tenant scope refers only to Platform API related beans provided by SDK.

**Review by kaczynskid-sag — COMMENTED** [review]
Much better effort than https://github.com/SoftwareAG/c8y-docs/pull/1509, but let's have a call and discuss this in detail @tsvet

**Review by jkot-softwareag — APPROVED** [review]

File: content/microservice-sdk/java-bundle/developing-microservice.md (hunk 14 above)
Comment (tsvet) [reply content/microservice-sdk/java-bundle/developing-microservice.md]: s. d9bc2ad8936e4cbb041129ef5cd5f8ade8e782bf

File: content/microservice-sdk/java-bundle/developing-microservice.md (hunk 15 above)
Comment (tsvet) [reply content/microservice-sdk/java-bundle/developing-microservice.md]: In d9bc2ad8936e4cbb041129ef5cd5f8ade8e782bf I have changed the flow to start with the ContextService and have corrected the sentence with setting the context.

File: content/microservice-sdk/java-bundle/developing-microservice.md (hunk 16 above)
Comment (tsvet) [reply content/microservice-sdk/java-bundle/developing-microservice.md]: s. d9bc2ad8936e4cbb041129ef5cd5f8ade8e782bf

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 17]
```diff
@@ -30,40 +30,98 @@ Annotation | Description
...
 
-`@UserScope` and `@TenantScope` at type level annotation indicate that a bean created from class will be created in the scope defined. The user scope implies using tenant platform user credentials for platform calls. The tenant scope implies using service user credentials.
+Each microservice has a service user which can be used for the interaction with the platform. The roles associated with this user are specified in the manifest. 
+Within the tenant scope, the credentials of this service user are used for the communication with the platform, while within the user scope the credentials of the authenticated user sending the request to the microservice are used.
 
-An example of injecting a bean into the tenant scope is available in the platform API module as follows:
+#### Setting the context
+You can explicitly set the context along with the credentials to be used through `ContextService`. To use the credentials of the user sending the request, use `ContextService<UserCredentials>`. Analogously, `ContextService<MicroserviceCredentials>` can be used for service user credentials. 
+Examples on how to use `ContextService` is given below.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:40]:
```suggestion
Examples on how to use `ContextService` are given below.
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 18]
```diff
@@ -30,40 +30,98 @@ Annotation | Description
...
+@Bean(name = "tenantEventApi")
+public EventApi eventApi (Platform platform) throws SDKException {
+  return platform.getEventApi();
+}  
+```
+
+By default, the Platform API related beans provided by the Microservice SDK are created in the tenant scope and use the service user to communicate with the platform.  
+
+There are predefined beans both in the `@TenantScope` and `@UserScope`. 
+The name of a bean in the tenant scope consists of the prefix `"tenant"` and the name of the respective API. Thus, to use the Event Api in the tenant scope, you can specify @Qualifier("tenantEventApi"), as shown in the example below. As the tenant scope is the default context for the created beans, the annotation can also be omitted. Therefore, the following two excerpts are equivalent and both suggest that the service user credentials will be used for the communication with the platform.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:85]:
```suggestion
The name of a bean in the tenant scope consists of the prefix `"tenant"` and the name of the respective API. Thus, to use the Event API in the tenant scope, you can specify @Qualifier("tenantEventApi"), as shown in the example below. As the tenant scope is the default context for the created beans, the annotation can also be omitted. Therefore, the following two excerpts are equivalent and both suggest that the service user credentials will be used for the communication with the platform. 
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 19]
```diff
@@ -30,40 +30,98 @@ Annotation | Description
...
 
-`@UserScope` and `@TenantScope` at type level annotation indicate that a bean created from class will be created in the scope defined. The user scope implies using tenant platform user credentials for platform calls. The tenant scope implies using service user credentials.
+Each microservice has a service user which can be used for the interaction with the platform. The roles associated with this user are specified in the manifest. 
+Within the tenant scope, the credentials of this service user are used for the communication with the platform, while within the user scope the credentials of the authenticated user sending the request to the microservice are used.
 
-An example of injecting a bean into the tenant scope is available in the platform API module as follows:
+#### Setting the context
+You can explicitly set the context along with the credentials to be used through `ContextService`. To use the credentials of the user sending the request, use `ContextService<UserCredentials>`. Analogously, `ContextService<MicroserviceCredentials>` can be used for service user credentials.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:39]:
```suggestion
You can explicitly set the context along with the credentials to be used through `ContextService`. To use the credentials of the user sending the request, use `ContextService<UserCredentials>`. Accordingly, `ContextService<MicroserviceCredentials>` can be used for service user credentials. 
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 20]
```diff
@@ -30,40 +30,98 @@ Annotation | Description
...
+
+public PagedEventCollectionRepresentation get10Events () {
+  return contextService.runWithinContext(contextService.getContext(), () -> eventApi.getEvents().get(10));
+}
 ```
+In this example, the events will be obtained using the credentials of the authenticated user.
 
-A sample utilization of the bean can be as follows:
 
+Analogously, the credentials of the service user will be utilized to retrieve the events in the example given below.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:55]:
```suggestion
In the second example, the credentials of the service user will be utilized to retrieve the events.
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 21]
```diff
@@ -30,40 +30,98 @@ Annotation | Description
...
+@Autowired
+private ContextService<UserCredentials> contextService;
+@Autowired
+private EventApi eventApi;
+
+public PagedEventCollectionRepresentation get10Events () {
+  return contextService.runWithinContext(contextService.getContext(), () -> eventApi.getEvents().get(10));
+}
 ```
+In this example, the events will be obtained using the credentials of the authenticated user.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:52]:
```suggestion
In this first example, the events are obtained using the credentials of the authenticated user.
```

File: content/microservice-sdk/java-bundle/developing-microservice.md [hunk 22]
```diff
@@ -30,40 +30,98 @@ Annotation | Description
...
+```java
+@Autowired
+private EventApi eventApi;
+```
+In both cases, beans within the tenant scope will be auto-wired. 
+
+#### User scope
+In certain situations the microservice should not use the service user credentials but the credentials of the user sending the request. 
+
+To create a bean in the user scope, specify `@UserScope`, as in the example below.
```
Comment (BeateRixen) [inline content/microservice-sdk/java-bundle/developing-microservice.md:102]:
```suggestion
To create a bean in the user scope, specify `@UserScope`:
```

**Review by BeateRixen — APPROVED** [review]

**Review by BeateRixen — APPROVED** [review]

**Review by jochenendres — APPROVED** [review]
